src/machines/java/com/enderio/machines/common/blockentity/ImpulseHopperBlockEntity.java

```diff
@@ -0,0 +1,112 @@
+
+    public ImpulseHopperBlockEntity(BlockEntityType<?> type, BlockPos worldPosition, BlockState blockState) {
+        super(MachineCapacitorKeys.IMPULSE_HOPPER_ENERGY_CAPACITY.get(),
+            MachineCapacitorKeys.IMPULSE_HOPPER_ENERGY_TRANSFER.get(),
+            MachineCapacitorKeys.IMPULSE_HOPPER_ENERGY_CONSUME.get(),
+            type, worldPosition, blockState);
+    }
+
+    @Override
+    public AbstractContainerMenu createMenu(int containerId, Inventory inventory, Player player) {
+        return new ImpulseHopperMenu(this, inventory, containerId);
+    }
+
+    @Override
+    public MachineInventoryLayout getInventoryLayout() {
+        return MachineInventoryLayout.builder(true).inputSlot(6).outputSlot(6).ghostSlot(6).build(); // first 6 input, second 6 output, third 6 ghost
+    }
+
+    @Override
+    public void serverTick() {
+        if(shouldActTick() && shouldPassItems()) {
+            passItems();
+        }
+        super.serverTick();
+    }
+
+    public boolean shouldActTick() {// TODO General tick method for power consuming devices?
+        return canAct() && level.getGameTime() % ticksForAction() == 0;
+    }
+
+    public int ticksForAction() {
+        return 20;
+    }
+
+    public boolean canPass(int slot) {
+        if (this.getInventory().getStackInSlot(slot).getItem().equals(this.getInventory().getStackInSlot(slot+6+6).getItem()) || this.getInventory().getStackInSlot(slot+6+6).isEmpty()) {
+            if (this.getInventory().getStackInSlot(slot).getCount() >= this.getInventory().getStackInSlot(slot+6+6).getCount()) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    public boolean canHold(int slot) {
+        return this.getInventory().getStackInSlot(slot+6).getCount() + this.getInventory().getStackInSlot(slot+6+6).getCount() <= 64;
+    }
+
+    public boolean shouldPassItems() {
+        int totalpower = 0;
+        for (int i = 0; i < 6; i++) {
+            if (canPass(i) && canHold(i)) {
+                totalpower += this.getInventory().getStackInSlot(i+6+6).getCount() * IMPULSE_HOPPER_POWER_USE_PER_ITEM;
+                continue;
+            }
+            return false;
+        }
+        if (this.getEnergyStorage().canConsumeEnergy(totalpower)) {
+            return true;
+        }
+        return false;
+    }
+
+    public void passItems() {
+        for (int i = 0; i < 6; i++) {
+            ItemStack stack = this.getInventory().getStackInSlot(i);
+            ItemStack ghost = this.getInventory().getStackInSlot(i+6+6);
+            ItemStack result = this.getInventory().getStackInSlot(i+6);
+            if (ghost.isEmpty()) {
+                continue;
+            }
+            if (result.isEmpty()) {
+                result = stack.copy();
+                result.setCount(ghost.getCount());
+            } else if (stack.is(result.getItem())) {
+                result.setCount(result.getCount() + ghost.getCount());
```

Comment:
Is this check not always true, because of the check in canPass? So the last else with a continue should never be able to run